heron/scheduler-core/src/java/com/twitter/heron/scheduler/SubmitterMain.java

```diff
@@ -389,30 +390,30 @@ public boolean submitTopology() {
       // TODO(mfu): timeout should read from config
       SchedulerStateManagerAdaptor adaptor = new SchedulerStateManagerAdaptor(statemgr, 5000);
 
-      boolean isValid = validateSubmit(adaptor, topology.getName());
+      validateSubmit(adaptor, topology.getName());
 
       // 2. Try to submit topology if valid
-      if (isValid) {
-        // invoke method to submit the topology
-        LOG.log(Level.FINE, "Topology {0} to be submitted", topology.getName());
-
-        // Firstly, try to upload necessary packages
-        packageURI = uploadPackage(uploader);
-        if (packageURI == null) {
-          LOG.severe("Failed to upload package.");
-          return false;
-        } else {
-          // Secondly, try to submit a topology
-          // build the runtime config
-          Config runtime = Config.newBuilder()
-              .putAll(LauncherUtils.getInstance().getPrimaryRuntime(topology, adaptor))
-              .put(Keys.topologyPackageUri(), packageURI)
-              .put(Keys.launcherClassInstance(), launcher)
-              .build();
-
-          isSuccessful = callLauncherRunner(runtime);
-        }
+      // invoke method to submit the topology
+      LOG.log(Level.FINE, "Topology {0} to be submitted", topology.getName());
+
+      // Firstly, try to upload necessary packages
+      packageURI = uploadPackage(uploader);
+      if (packageURI == null) {
+        throw new TopologySubmissionException("Failed to upload package");
```

Comment:
Actually looks like you already throw UploaderException below, so we should be able to remove the null handling.